Домашнее задание №2, первая попытка сдачи.

IgushkinV opened the pull request:
В этот раз все операции с гитом (ветвление, коммит и пуш) делал из IDEA. Только пуллреквест создан с гитхаба.

IgushkinV committed:
Домашнее задание №2, первая попытка сдачи.

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/Cache.java:
```diff
@@ -0,0 +1,147 @@
+
+    private int capacityOfCache;
+    private CacheElement[] cache;
+    private int indexOfMostRightNotNullCacheElement;
```
Если шутить, то можно было бы назвать еще длинее)
А если серьезно, то чем lastIndex или еще как не угодил?

IgushkinV replied:
lastIndex - это путает, потому что не подходит просто последний индекс в массиве. lastUsedIndex - индекс, к которому обращались последним. И так далее) В итоге ничего короче не придумал... Когда код написан, уже можно и lastIndex назвать.

IgushkinV replied:
Добавил комментарий на поле.

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/Cache.java:
```diff
@@ -0,0 +1,147 @@
+     * @param elementToAdd элемент для добавления в кэш
+     * @param index уникальный для каждого элемента в кэше индекс
+     */
+    public void add(CacheElement<T> elementToAdd, int index) {
```
Хорошо конечно что пытаешься дать смысл названиям, но слишком длинно выходит, может newElement или просто element?
И почему тут на добавление дается CacheElement? Этот класс используется внутри Cache, снаружи зачем его использовать?

IgushkinV replied:
Просто element не подошло, т.к. их слишком много в задании. Начали путать, кто из них кто, поэтому решил изменить имя аргумента. newElement тоже хотел, но когда начинаешь набирать new, то IDE подсказывает создание нового объекта. В итоге получилось всего на 2 буквы длиннее, чем newElement :)

IgushkinV replied:
По второму замечанию да, нелогично вышло. Исправил.

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/Cache.java:
```diff
@@ -0,0 +1,147 @@
+     * Если происходит запрос на удаление из пустого кэша, то просто выходит из метода.
+     * @param element элемент для удаления
+     */
+    public void delete(CacheElement<T> element) {
```
Опять CacheElement

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/Cache.java:
```diff
@@ -0,0 +1,147 @@
+            cache[indexOfMostRightNotNullCacheElement].index = index;
+            return;
+        }
+        if (indexOfMostRightNotNullCacheElement == capacityOfCache - 1) { //Если кэш заполнен полностью
```
Давай вынесем в отдельный метод сдвиг и будем его переиспользовать?

IgushkinV replied:
Сделано.

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/Cache.java:
```diff
@@ -0,0 +1,147 @@
+            }
+        }
+        if (numForDelete >= 0) {
+            for (int i = numForDelete; i < capacityOfCache - 1; i++) {
```
Давай переиспользуем метод что ты вынесешь из метода add?

IgushkinV replied:
Сделано.

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/Cache.java:
```diff
@@ -0,0 +1,147 @@
+        }
+        for (int i = 0; i <= indexOfMostRightNotNullCacheElement; i++) {
+            if (cache[i] != null) {
+                if (cache[i].index == index) {
```
Для чего тут 2 условия, может совместим в одно через && ?

IgushkinV replied:
Не заметил сначала. Исправлено в коммите "Исправления №2".

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/Cache.java:
```diff
@@ -0,0 +1,147 @@
+        for (int i = 0; i <= indexOfMostRightNotNullCacheElement; i++) {
+            if (cache[i].index == index) {
+                returnElement = cache[i];
+                for (int j = i; j <= indexOfMostRightNotNullCacheElement - 1; j++) {
```
Еще одно переиспользование сдвига можно сделать

IgushkinV replied:
Сделано.

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/CacheElement.java:
```diff
@@ -0,0 +1,10 @@
+package Igushkin.Lesson2;
+
+public class CacheElement<T> {
+    T element;
```
модификаторы доступа?

IgushkinV replied:
Не заметил комментарий сначала. Исправил в коммите "Исправление замечаний к ДЗ №2".
Добавил полям модификаторы доступа  private и геттеры и сеттеры на поля.

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/CacheElement.java:
```diff
@@ -0,0 +1,10 @@
+    T element;
+    int index;
+
+    public CacheElement(T element) {
```
А чего индекс не добавил?

IgushkinV replied:
Сделано. Создание новых объектов в коде исправил.

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/CacheElement.java:
```diff
@@ -0,0 +1,10 @@
+
+    public CacheElement(T element) {
+        this.element = element;
+    }
```
equals?

IgushkinV replied:
Добавил переопределение equals и hashCode

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/Storage.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    @SuppressWarnings("unchecked")
+    public void add (T element) {
+        if (indexOfMostRightNotNullStorageElem == storageCapacity - 1) {
```
Давай вынесем увеличение размера в отдельный метод и переиспользуем тут?

IgushkinV replied:
Сделано. Добавил константу FACTOR.

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/Storage.java:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * Очищает кэш.
+     */
+    public void clear() {
```
В задачке вроде не описано требование, давай тут еще очистим наш массив в storage?

IgushkinV replied:
Сделано.

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/Storage.java:
```diff
@@ -0,0 +1,103 @@
+        if (cache.isPresent(index)) {
+            return cache.get(index).element;
+        }
+        cache.add(new CacheElement<>(storage[index]), index);
```
Тут мы должен элемент параметризованный передавать, а не CacheElement

IgushkinV replied:
Исправлено.

IgushkinV committed:
Исправление замечаний к ДЗ.

IgushkinV committed:
Исправление замечаний к ДЗ №2.

IgushkinV committed:
Исправление замечаний к ДЗ №3.
Использование  getElement() вместо обращения напрямую к полю.

IgushkinV committed:
Исправление замечаний к ДЗ №4.
Исправил второй конструктор Storage - неправильно устанавливался lastIndex.
Исправил конструктор в Cache - была ошибка java.lang.ClassCastException: class [Ljava.lang.Object; cannot be cast to class [LIgushkin.Lesson2.CacheElement;
Исправил delete(T element).
